precompile/erc20/approve.go

```diff
@@ -156,20 +172,60 @@ func createAuthorization(ctx sdk.Context, grantee, granter common.Address, amoun
 	return authzkeeper.SaveGrant(ctx, grantee.Bytes(), granter.Bytes(), authorization, &expiration)
 }
 
-func updateAuthorization(ctx sdk.Context, grantee, granter common.Address, amount *big.Int, authorization authz.Authorization, expiration *time.Time, authzkeeper authzkeeper.Keeper) error {
+func updateAuthorization(
+	ctx sdk.Context,
+	denom string,
+	grantee, granter common.Address,
+	amount *big.Int,
+	authorization authz.Authorization,
+	expiration *time.Time,
+	authzkeeper authzkeeper.Keeper,
+) error {
 	sendAuthz, ok := authorization.(*banktypes.SendAuthorization)
 	if !ok {
 		return fmt.Errorf("unknown authorization type")
 	}
 
-	sdkAmount, err := precompile.TypesConverter.BigInt.ToSDK(amount)
+	// Caller ensures targetAmount is >= 0.
+	targetAmount, err := precompile.TypesConverter.BigInt.ToSDK(amount)
 	if err != nil {
 		return fmt.Errorf("failed to convert amount: [%w]", err)
 	}
 
-	sendAuthz.SpendLimit[0] = sdk.Coin{Denom: evm.DefaultEVMDenom, Amount: sdkAmount}
+	// Sort the spend limit to ensure the isSorted invariant is maintained for all
+	// below operations made on the spend limit. This is not strictly required if
+	// the authorization is managed only by this precompile function
+	// (the spend limit is sorted upon update) but may save us in case the
+	// authorization is ever modified somewhere else.
+	sendAuthz.SpendLimit = sendAuthz.SpendLimit.Sort()
+
+	currentAmount := sendAuthz.SpendLimit.AmountOfNoDenomValidation(denom)
+	deltaAmount := targetAmount.Sub(currentAmount)
+
+	switch deltaAmount.Sign() {
```

Comment:
Thoughts on simplifying it a bit?
```
  if deltaAmount.Sign() == 0 {
		return nil
	}

	deltaCoins := sdk.Coins{{Denom: denom, Amount: deltaAmount}}
	sendAuthz.SpendLimit = sendAuthz.SpendLimit.Add(deltaCoins...)
```

In case `delta < 0` we do `deltaAmount.Neg()`, so I think without `Neg()` we can use `Add()` which will result in subtractig the amounts. Up to you.

Reply:
I was thinking about it but not entirely sure how Cosmos `Coins` handle negatives and whether that works as expected in all cases. Given the time pressure I went the safer way.